src/box/tuple_constraint_fkey.c

```diff
@@ -0,0 +1,349 @@
+			struct index *findex = fspace->index[findex_no];
+			struct key_def *fkey_def = findex->def->key_def;
+			if (tuple_compare(deleted_tuple, HINT_NONE,
+					  replaced_with_tuple, HINT_NONE,
+					  fkey_def) == 0)
+				return 0;
+		}
+	}
+
+	if (constr->fkey->local_index < 0) {
+		foreign_key_integrity_failed(constr, "index was not found");
+		return -1;
+	}
+
+	struct index *index = constr->space->index[constr->fkey->local_index];
+	struct key_def *key_def = index->def->key_def;
+	uint32_t part_count = constr->fkey->field_count;
+	assert(constr->fkey->field_count == key_def->part_count);
+
+	const char *key = tuple_field(deleted_tuple, foreign_field_no);
+	if (key == NULL || mp_typeof(*key) == MP_NIL) {
+		/* No field - nobody can be bound to it.*/
+		return 0;
+	}
+
+	const char *unused;
+	if (key_validate_parts(key_def, key, part_count, false, &unused) != 0) {
+		foreign_key_integrity_failed(constr, "wrong key type");
+		return -1;
+	}
+
+	struct tuple *found_tuple;
+	if (index->def->opts.is_unique ?
+	    index_get(index, key, part_count, &found_tuple) :
+	    index_min(index, key, part_count, &found_tuple) != 0)
+		return -1;
+	if (found_tuple != NULL) {
+		foreign_key_integrity_failed(constr, "tuple is referenced");
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Destructor that unpins space from space_cache.
+ */
+static void
+tuple_constraint_fkey_unpin(struct tuple_constraint *constr)
+{
+	assert(constr->destroy == tuple_constraint_fkey_unpin);
+	space_cache_unpin(&constr->space_cache_holder);
+	constr->check = tuple_constraint_noop_check;
+	constr->destroy = tuple_constraint_noop_destructor;
+	constr->space = NULL;
+}
+
+/**
+ * Find and set foreign_field amd foreign_index fkey member of @a constraint.
+ * If something was not found - foreign_index is set to -1.
+ */
+static void
+tuple_constraint_fkey_update_foreign(struct tuple_constraint *constraint)
+{
+	struct space *space = constraint->space_cache_holder.space;
+	constraint->fkey->foreign_index = -1;
+	assert(constraint->fkey->field_count == 1);
+	constraint->fkey->data[0].foreign_field_no =
+		find_field_no_by_def(space, &constraint->def.fkey.field);
+	if (constraint->fkey->data[0].foreign_field_no >= 0)
+		fkey_find_foreign_index(constraint);
+}
+
+/**
+ * Callback that is called when a space that is pinned by a constraint is
+ * replaced in space cache.
+ */
+static void
+tuple_constraint_fkey_space_cache_on_replace(struct space_cache_holder *holder,
+					     struct space *old_space)
+{
+	(void)old_space;
+	struct tuple_constraint *constr =
+		container_of(holder, struct tuple_constraint,
+			     space_cache_holder);
+	tuple_constraint_fkey_update_foreign(constr);
+}
+
+int
+tuple_constraint_fkey_init(struct tuple_constraint *constr,
+			   struct space *space, int32_t field_no)
+{
+	assert(constr->def.type == CONSTR_FKEY);
+	constr->space = space;
+	constr->fkey->data[0].local_field_no = field_no;
+	fkey_find_local_index(constr);
```

Comment:
Again, the function name implies that it finds and returns an index (which it does AFAICS), but you don't use the return value for some reason. This looks confusing.